Zend/zend_vm_def.h

```diff
@@ -9597,46 +9597,89 @@ ZEND_VM_C_LABEL(try_again):
 	}
 }
 
-ZEND_VM_HANDLER(204, ZEND_FRAMELESS_ICALL_0, UNUSED, UNUSED)
+ZEND_VM_HANDLER(204, ZEND_FRAMELESS_ICALL_0, UNUSED, UNUSED, SPEC(OBSERVER))
 {
 	USE_OPLINE
 	SAVE_OPLINE();
+
+#if !ZEND_VM_SPEC || ZEND_OBSERVER_ENABLED
+	if (ZEND_OBSERVER_ENABLED) {
+		zend_function *fbc = ZEND_FLF_FUNC(opline);
+		if (UNEXPECTED(zend_observer_handler_is_unobserved(ZEND_OBSERVER_DATA(fbc)) == false)) {
+			zend_execute_data *call = _zend_vm_stack_push_call_frame_ex(zend_vm_calc_used_stack(0, fbc), ZEND_CALL_NESTED_FUNCTION, fbc, 0, NULL);
+			call->prev_execute_data = EX(call);
+			EX(call) = call;
+			ZEND_VM_DISPATCH_TO_HANDLER(ZEND_DO_ICALL);
+		}
+	}
+#endif
 	zend_frameless_function_0 function = (zend_frameless_function_0)ZEND_FLF_HANDLER(opline);
-	ZVAL_NULL(EX_VAR(opline->result.var));
+	zval *result = EX_VAR(opline->result.var);
+	ZVAL_NULL(result);
 	function(EX_VAR(opline->result.var));
 	ZEND_VM_NEXT_OPCODE_CHECK_EXCEPTION();
 }
 
-ZEND_VM_HANDLER(205, ZEND_FRAMELESS_ICALL_1, ANY, UNUSED)
+ZEND_VM_HANDLER(205, ZEND_FRAMELESS_ICALL_1, ANY, UNUSED, SPEC(OBSERVER))
 {
 	USE_OPLINE
 	SAVE_OPLINE();
 	zend_frameless_function_1 function = (zend_frameless_function_1)ZEND_FLF_HANDLER(opline);
-	ZVAL_NULL(EX_VAR(opline->result.var));
+	zval *result = EX_VAR(opline->result.var);
+	ZVAL_NULL(result);
 	zval *arg1 = GET_OP1_ZVAL_PTR_DEREF(BP_VAR_R);
-	if (EG(exception)) {
+	if (UNEXPECTED(EG(exception) != NULL)) {
 		FREE_OP1();
 		HANDLE_EXCEPTION();
 	}
-	function(EX_VAR(opline->result.var), arg1);
+#if !ZEND_VM_SPEC || ZEND_OBSERVER_ENABLED
+	if (ZEND_OBSERVER_ENABLED) {
+		zend_function *fbc = ZEND_FLF_FUNC(opline);
+		if (UNEXPECTED(zend_observer_handler_is_unobserved(ZEND_OBSERVER_DATA(fbc)) == false)) {
+			zend_execute_data *call = _zend_vm_stack_push_call_frame_ex(zend_vm_calc_used_stack(1, fbc), ZEND_CALL_NESTED_FUNCTION, fbc, 1, NULL);
+			call->prev_execute_data = EX(call);
+			ZVAL_COPY(ZEND_CALL_VAR_NUM(call, 0), arg1);
+			FREE_OP1();
```

Comment:
In general, COPY with the following FREE may be optimized to avoid refcounting.

Reply:
How? Do we have a proper macro for that, or do you mean having specialized #if SPEC?

Reply:
> How? Do we have a proper macro for that, or do you mean having specialized #if SPEC?

This specialization logic expanded in many VM places. e.g in ZEND_VAL/VAR. Also see ``zend_copy_to_variable()``.
Probably, it doesn't make sense to do this for unspecialised observer handlers.

Reply:
Right, these aren't specialized. So yes, I agree.